### PR TITLE
changed recommend

### PR DESCRIPTION
**Pull request by spirinamayya**

### CODE REVIEW

File: rectools/models/sasrec.py [hunk 1]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
-            item_features=item_features,
-            top_k=k,
-            candidate_items=items_to_recommend,
+        session_embs = []
+        device = self.model.get_model_device()
+        self.model.item_model.to_device(device)
+        with torch.no_grad():
+            for x_batch in tqdm.tqdm(recommend_dataloader):
+                x_batch = x_batch.to(device)  # [batch_size, session_maxlen]
+                session_batch = self.model.predict(x_batch)  # [batch_size, n_items]
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
                session_batch = self.model.predict(x_batch)  # [batch_size, factors]
```

File: rectools/models/sasrec.py [hunk 2]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
+                session_batch = session_batch.detach().cpu().numpy()
+                session_embs.append(session_batch)
+        session_embs = np.concatenate(session_embs, axis=0)  # [n_users, n_items]
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
        session_embs = np.concatenate(session_embs, axis=0)  # [n_rec_users, factors]
```

File: rectools/models/sasrec.py [hunk 3]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
+
+        item_embs = self.model.item_model.get_all_embeddings()
+        item_embs = item_embs[candidate_items_inds]
+        item_embs = np.array(item_embs.detach().cpu().numpy())
+
+        user_map = IdMap.from_values(user_ids)
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
        user_map = IdMap.from_values(user_ids) # 1x internal -> rec_user_internal

File: rectools/models/sasrec.py [hunk 4]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
+        filter_viewed: bool,
+        sorted_item_ids_to_recommend: tp.Optional[InternalIdsArray],
+    ) -> InternalRecoTriplet:
+        sorted_item_ids_to_recommend = dataset.item_id_map.convert_to_external(sorted_item_ids_to_recommend)
+        sorted_item_ids_to_recommend = pd.Series(
+            sorted(self.item_id_map.convert_to_internal(sorted_item_ids_to_recommend))
+        )
         train = dataset.get_raw_interactions()
+        mask = pd.Series(dataset.user_id_map.convert_to_internal(train[Columns.User])).isin(user_ids)
+        rec_df = train[mask]
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
        rec_df = train[mask] # external
```

File: rectools/models/sasrec.py [hunk 5]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
+        rec_df[Columns.Item] = self.item_id_map.convert_to_internal(rec_df[Columns.Item])
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
        rec_df[Columns.Item] = self.item_id_map.convert_to_internal(rec_df[Columns.Item]) # model internal
```

File: rectools/models/sasrec.py [hunk 6]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
-        assume_external_ids: bool = True,
-    ) -> pd.DataFrame:
-        """TODO"""
+        filter_viewed: bool,
+        sorted_item_ids_to_recommend: tp.Optional[InternalIdsArray],
+    ) -> InternalRecoTriplet:
+        sorted_item_ids_to_recommend = dataset.item_id_map.convert_to_external(sorted_item_ids_to_recommend)
+        sorted_item_ids_to_recommend = pd.Series(
+            sorted(self.item_id_map.convert_to_internal(sorted_item_ids_to_recommend))
+        )
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
        ) # model internal
```

File: rectools/models/sasrec.py [hunk 7]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
+        mask = pd.Series(dataset.user_id_map.convert_to_internal(train[Columns.User])).isin(user_ids)
+        rec_df = train[mask]
+        rec_df[Columns.Item] = self.item_id_map.convert_to_internal(rec_df[Columns.Item])
+
+        items = pd.Series(self.item_id_map.get_sorted_internal()[1:])
+        if sorted_item_ids_to_recommend is None:
+            candidate_items = items
+        else:
+            candidate_items = sorted_item_ids_to_recommend[sorted_item_ids_to_recommend.isin(items)]
+        candidate_items_inds = torch.LongTensor(candidate_items)
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
        candidate_items_inds = torch.LongTensor(candidate_items) # model internal
```

File: rectools/models/sasrec.py (hunk 5 above)
Comment (blondered) [inline rectools/models/sasrec.py]: Поправь меня плиз, но я не вижу:
1. где мы фильтруем из rec_df неизвестные айтемы
2. где мы гарантируем последовательность айдишников юзеров
3. кажется, что sparse матрица для filter_viewed имеет несовпадающие айдишники с теми, что подаются в ImplicitRanker

File: rectools/models/sasrec.py [hunk 8]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
+                session_embs.append(session_batch)
+        session_embs = np.concatenate(session_embs, axis=0)  # [n_users, n_items]
+
+        item_embs = self.model.item_model.get_all_embeddings()
+        item_embs = item_embs[candidate_items_inds]
+        item_embs = np.array(item_embs.detach().cpu().numpy())
+
+        user_map = IdMap.from_values(user_ids)
+        rec_df[Columns.User] = dataset.user_id_map.convert_to_internal(rec_df[Columns.User])
+        rec_df[Columns.User] = user_map.convert_to_internal(rec_df[Columns.User])
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
        rec_df[Columns.User] = user_map.convert_to_internal(rec_df[Columns.User]) # rec_user_internal
```

File: rectools/models/sasrec.py [hunk 9]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
+        if filter_viewed:
+            values = np.ones(len(rec_df))
+            ui_csr_for_filter = sparse.csr_matrix(
```
Comment (blondered) [inline rectools/models/sasrec.py:258]:
```suggestion
            ui_csr_for_filter = sparse.csr_matrix( # [n_rec_users x n_train_items]
```

File: rectools/models/sasrec.py [hunk 10]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
+                (
+                    values.astype(np.float32),
+                    (
+                        rec_df[Columns.User].values,
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
                        rec_df[Columns.User].values, # rec_user_internal
```

File: rectools/models/sasrec.py [hunk 11]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
+                        rec_df[Columns.Item].values - 1,
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
                        rec_df[Columns.Item].values - 1, # model internal
```

File: rectools/models/sasrec.py [hunk 12]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
+                    ),
+                ),
+            )
+        else:
+            ui_csr_for_filter = None
+
+        ranker = ImplicitRanker(Distance.DOT, session_embs, item_embs)
+        all_target_ids, all_reco_ids, all_scores = ranker.rank(
+            subject_ids=np.arange(len(user_ids)),
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
            subject_ids=np.arange(len(user_ids)), # rec_user_internal
```

File: rectools/models/sasrec.py [hunk 13]
```diff
@@ -107,34 +113,113 @@ def _fit(
...
+                    (
+                        rec_df[Columns.User].values,
+                        rec_df[Columns.Item].values - 1,
+                    ),
+                ),
+            )
+        else:
+            ui_csr_for_filter = None
+
+        ranker = ImplicitRanker(Distance.DOT, session_embs, item_embs)
```
Comment (blondered) [inline rectools/models/sasrec.py]:
```suggestion
        ranker = ImplicitRanker(
            Distance.DOT, 
            session_embs, # [n_rec_users x factors]
            item_embs # [n_candidates x factors]
        )
```
